codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/RestJsonProtocolGenerator.java

```diff
@@ -44,55 +61,305 @@
     }
 
     @Override
-    protected void generateDocumentBodyShapeSerializers(GenerationContext context, Set<Shape> shapes) {
+    protected void generateOperationDocumentSerializer(
+            GenerationContext context, OperationShape operation
+    ) {
+        Model model = context.getModel();
+
+        HttpBindingIndex bindingIndex = model.getKnowledge(HttpBindingIndex.class);
+
+        Set<MemberShape> bindingMap = bindingIndex.getRequestBindings(operation).values().stream()
+                .filter(binding -> binding.getLocation() == HttpBinding.Location.DOCUMENT)
+                .map(HttpBinding::getMember)
+                .collect(Collectors.toSet());
+
+        if (bindingMap.size() == 0) {
+            return;
+        }
+
+        Shape inputShape = model.expectShape(operation.getInput()
+                .orElseThrow(() -> new CodegenException("input shapre missing for operation " + operation.getId())));
+
+        GoWriter writer = context.getWriter();
+
+        writeJsonShapeSerializerFunction(writer, model, context.getSymbolProvider(), inputShape,
+                bindingMap::contains);
+        writer.write("");
     }
 
-    @Override
-    protected void generateDocumentBodyShapeDeserializers(GenerationContext context, Set<Shape> shapes) {
+    private void writeJsonShapeSerializerFunction(
+            GoWriter writer,
+            Model model,
+            SymbolProvider symbolProvider,
+            Shape inputShape,
+            Function<MemberShape, Boolean> filterMemberShapes
+    ) {
+        Symbol jsonEncoder = SymbolUtils.createValueSymbolBuilder("Value", GoDependency.AWS_JSON_PROTOCOL)
+                .build();
+
+        Symbol inputSymbol = symbolProvider.toSymbol(inputShape);
+
+        String functionName = ProtocolGenerator.getDocumentSerializerFunctionName(inputShape, getProtocolName());
+
+        writer.addUseImports(SymbolUtils.createValueSymbolBuilder(null, GoDependency.FMT).build());
+        writer.addUseImports(inputSymbol);
+        writer.addUseImports(jsonEncoder);
```

Comment:
Had the same comments on the other PR, but copying here: symbols will have their deps be imported automatically when you write them out with either the T or P formaters and `GoDependency`s can be imported directly.